File: examples/09-ipc/polyglot/java/src/main/java/io/orionmesh/demo/Pub.java
package io.orionmesh.demo;

import io.nats.client.Connection;
import io.nats.client.Nats;

import java.time.Instant;
import java.time.ZoneOffset;
import java.time.format.DateTimeFormatter;
import java.util.Map;

/**
 * Java NATS publisher — interoperates with the Rust + Python demos.
 *
 * Args (all optional, dash-prefixed style for parity with the Rust/Python demos):
 *   --nats-url nats://127.0.0.1:4222
 *   --subject  orion.demo.ipc
 *   --interval 1.0
 *   --label    java
 *
 * Reads ORION_REPLICA_INDEX env var (set by the OrionMesh agent when launched
 * as one of N replicas) for the default label.
 */
public class Pub {
    public static void main(String[] argv) throws Exception {
        Map<String, String> args = parseArgs(argv);
        String url = args.getOrDefault("--nats-url",
                System.getenv().getOrDefault("NATS_URL", "nats://127.0.0.1:4222"));
        String subject = args.getOrDefault("--subject", "orion.demo.ipc");
        double interval = Double.parseDouble(args.getOrDefault("--interval", "1.0"));
        String label = args.getOrDefault("--label", labelFromEnv());

        System.out.printf("[java-pub:%s] connecting to %s -> %s%n", label, url, subject);
        try (Connection nc = Nats.connect(url)) {
            System.out.printf("[java-pub:%s] connected%n", label);
            long i = 0;
            DateTimeFormatter fmt = DateTimeFormatter.ofPattern("HH:mm:ss.SSS").withZone(ZoneOffset.UTC);
            while (true) {
                i++;
                String ts = fmt.format(Instant.now());
                String line = String.format("tick %d from %s at %s", i, label, ts);
                nc.publish(subject, line.getBytes());
                System.out.printf("[java-pub:%s] sent: %s%n", label, line);
                Thread.sleep((long) (interval * 1000));
            }
        }
    }

    private static String labelFromEnv() {
        String idx = System.getenv("ORION_REPLICA_INDEX");
        return idx != null ? "r" + idx : "java";
    }

    private static Map<String, String> parseArgs(String[] argv) {
        var out = new java.util.HashMap<String, String>();
        for (int i = 0; i + 1 < argv.length; i += 2) {
            out.put(argv[i], argv[i + 1]);
        }
        return out;
    }
}

File: examples/09-ipc/polyglot/python/pub.py
#!/usr/bin/env python3
"""Python NATS publisher — interoperates with the Rust + Java demos.

Publishes "tick N from <label> at HH:MM:SS.mmm" to a NATS subject every
--interval seconds. Reads ORION_REPLICA_INDEX (set by the OrionMesh agent
when launched as one of N replicas) to self-identify.

Setup:
  python3 -m venv .venv
  . .venv/bin/activate
  pip install nats-py
"""
from __future__ import annotations

import argparse
import asyncio
import os
import sys
from datetime import datetime, timezone

try:
    import nats
except ImportError:
    sys.stderr.write("nats-py not installed. Run: pip install nats-py\n")
    sys.exit(1)


async def main():
    p = argparse.ArgumentParser()
    p.add_argument("--nats-url", default=os.environ.get("NATS_URL", "nats://127.0.0.1:4222"))
    p.add_argument("--subject", default="orion.demo.ipc")
    p.add_argument("--interval", type=float, default=1.0)
    p.add_argument("--label", default=None)
    args = p.parse_args()

    label = args.label or f"r{os.environ.get('ORION_REPLICA_INDEX')}" if os.environ.get("ORION_REPLICA_INDEX") else (args.label or "py")
    print(f"[py-pub:{label}] connecting to {args.nats_url} -> {args.subject}", flush=True)
    nc = await nats.connect(args.nats_url)
    print(f"[py-pub:{label}] connected", flush=True)
    i = 0
    try:
        while True:
            i += 1
            ts = datetime.now(timezone.utc).strftime("%H:%M:%S.") + f"{datetime.now().microsecond // 1000:03d}"
            line = f"tick {i} from {label} at {ts}"
            await nc.publish(args.subject, line.encode())
            print(f"[py-pub:{label}] sent: {line}", flush=True)
            await asyncio.sleep(args.interval)
    finally:
        await nc.drain()


if __name__ == "__main__":
    asyncio.run(main())
